modules/util/src/main/scala/gpp/util/Zipper.scala

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2016-2020 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gpp.util
+
+import cats._
+import cats.implicits._
+import cats.data.NonEmptyList
+import monocle.{ Prism, Traversal }
+import monocle.macros.Lenses
+
+/**
+  * Minimal zipper based on scalaz's implementation
+  * This is only meant for small collections. performance has not been optimized
+  */
+@Lenses
+final case class Zipper[A](lefts: List[A], focus: A, rights: List[A]) {
+
+  /**
+    * Modify the focus
+    */
+  def modify(f: A => A): Zipper[A] = copy(lefts, f(focus), rights)
+
+  /**
+    * Modify the focus
+    */
+  def modifyP(p: Prism[A, A]): Zipper[A] =
+    p.getOption(focus).map(f => copy(lefts, f, rights)).getOrElse(this)
+
+  /**
+    * Find and element and focus if successful
+    */
+  def findFocusP(p: PartialFunction[A, Boolean]): Option[Zipper[A]] =
+    findFocus(p.lift.andThen(_.getOrElse(false)))
+
+  /**
+    * How many items are in the zipper
+    */
+  def length: Int = lefts.length + 1 + rights.length
+
+  /**
+    * Find and element and focus if successful
+    */
+  def findFocus(p: A => Boolean): Option[Zipper[A]] =
+    if (p(focus)) this.some
+    else {
+      val indexLeft  = lefts.indexWhere(p)
+      val indexRight = rights.indexWhere(p)
+      if (indexLeft === -1 && indexRight === -1)
+        none
+      else if (indexLeft >= 0)
+        (lefts.splitAt(indexLeft) match {
+          case (Nil, h :: t)      =>
+            Zipper(Nil, h, t ::: focus :: rights)
+          case (x :: Nil, i :: l) =>
+            Zipper(List(x), i, l ::: focus :: rights)
+          case (x, i :: l)        =>
+            Zipper(x, i, (focus :: l.reverse).reverse ::: rights)
+          case _                  =>
+            this
+        }).some
+      else
+        (rights.splitAt(indexRight) match {
+          case (Nil, List(y)) =>
+            Zipper((focus :: lefts.reverse).reverse, y, Nil)
+          case (x, h :: t)    =>
+            Zipper((focus :: lefts.reverse).reverse ::: x, h, t)
+          case _              =>
+            this
+        }).some
+    }
+
+  def exists(p: A => Boolean): Boolean =
+    if (p(focus))
+      true
+    else
+      lefts.exists(p) || rights.exists(p)
+
+  def previous: Option[Zipper[A]] =
+    lefts match {
+      case Nil      => none
+      case x :: Nil => Zipper(Nil, x, focus :: rights).some
+      case _        => Zipper(lefts.init, lefts.last, focus :: rights).some
```

Comment:
This works but the idea with a list `Zipper`, I thought, was to store the `lefts` in reverse order so that you can move to the left or right in constant time.  For example, that way the last case could be:

```scala
case h :: t => Zipper(t, h, focus :: rights).some
```

and moving to the right would be

```scala
case h :: t => Zipper(focus :: lefts, h, t).some
```